### PR TITLE
gl_engine: make GlRenderTask generic with uniform block

### PR DESCRIPTION
**Pull request by RuiwenTang**
This PR does the following and completely rewrites  `GLRenderTask` :
* Use uniform block to pack all color information
* Move the actual gl draw call into GlRenderer::sync function, so all data is been uploaded into GPU
* Make GlRenderTask simple and generic for direct gl draw

### CODE REVIEW

File: src/renderer/gl_engine/tvgGlGeometry.cpp [hunk 1]
```diff
@@ -34,39 +35,22 @@ GlGeometry::~GlGeometry()
...
-    mStrokeCount = 0;
-
     if (flag & (RenderUpdateFlag::Color | RenderUpdateFlag::Gradient | RenderUpdateFlag::Transform)) {
-        Array<float> vertex;
-        Array<uint32_t> index;
+        mFillVertex.clear();
+        mFillIndex.clear();
 
-        tvg::Tessellator tess{&vertex, &index};
+        tvg::Tessellator tess{&mFillVertex, &mFillIndex};
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlGeometry.cpp:44]: let's remove tvg prefix

File: src/renderer/gl_engine/tvgGlGeometry.cpp [hunk 2]
```diff
@@ -34,39 +35,22 @@ GlGeometry::~GlGeometry()
...
     }
 
     if (flag & (RenderUpdateFlag::Stroke | RenderUpdateFlag::Transform)) {
-        Array<float> vertex;
-        Array<uint32_t> index;
+        mStrokeVertex.clear();
+        mStrokeIndex.clear();
 
-        tvg::Stroker stroke{&vertex, &index};
+        tvg::Stroker stroke{&mStrokeVertex, &mStrokeIndex};
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlGeometry.cpp:52]: --tvg

File: src/renderer/gl_engine/tvgGlGeometry.cpp [hunk 3]
```diff
@@ -79,22 +63,34 @@ void GlGeometry::disableVertex(uint32_t location)
...
+        return false;
     }
 
+    Array<float>* vertexBuffer = nullptr;
+    Array<uint32_t>* indexBuffer = nullptr;
 
-    uint32_t vertexOffset = (flag == RenderUpdateFlag::Stroke) ? mStrokeVertexOffset : mFillVertexOffset;
-    uint32_t indexOffset = (flag == RenderUpdateFlag::Stroke) ? mStrokeIndexOffset : mFillIndexOffset;
-    uint32_t count = (flag == RenderUpdateFlag::Stroke) ? mStrokeCount : mFillCount;
+    if (flag == RenderUpdateFlag::Stroke) {
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlGeometry.cpp:76]: Not this??   (flag & RenderUpdateFlag::Stroke)

File: src/renderer/gl_engine/tvgGlGeometry.h [hunk 4]
```diff
@@ -180,26 +180,25 @@ class GlPoint
...
     float* getTransforMatrix();
 
 private:
-    uint32_t mFillVertexOffset;
-    uint32_t mFillIndexOffset;
-    uint32_t mFillCount;
-    uint32_t mStrokeVertexOffset;
-    uint32_t mStrokeIndexOffset;
-    uint32_t mStrokeCount;
+    Array<float> mFillVertex = {};
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlGeometry.h:198]: For consistency, I'd like to suggest you to remove "m" from the member variables name.

File: src/renderer/gl_engine/tvgGlGpuBuffer.cpp [hunk 5]
```diff
@@ -60,6 +72,12 @@ void GlGpuBuffer::unbind(Target target)
     GL_CHECK(glBindBuffer(static_cast<uint32_t>(target), 0));
 }
 
+GlStageBuffer::GlStageBuffer()
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlGpuBuffer.cpp:75]: Suggested one line:  "GlStageBuffer::GlStageBuffer() : mVao(0), mGpuBuffer(new GlGpuBuffer)"

File: src/renderer/gl_engine/tvgGlGeometry.cpp [hunk 6]
```diff
@@ -79,22 +63,34 @@ void GlGeometry::disableVertex(uint32_t location)
 }
 
 
-void GlGeometry::draw(const uint32_t location, RenderUpdateFlag flag)
+bool GlGeometry::draw(GlRenderTask* task, GlStageBuffer* gpuBuffer, RenderUpdateFlag flag)
 {
 
     if (flag == RenderUpdateFlag::None) {
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlGeometry.cpp:69]: Suggested one line: if (xx) return false;

File: src/renderer/gl_engine/tvgGlGpuBuffer.cpp [hunk 7]
```diff
@@ -29,6 +29,18 @@
 /* Internal Class Implementation                                        */
 /************************************************************************/
 
+static GLint _getGpuBufferAlign() 
+{
+    static GLint offset = 0;
+
+    if (!offset)
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlGpuBuffer.cpp:36]: Suggested one line: if (!offset) GL_CHECK(......);

File: src/renderer/gl_engine/tvgGlGpuBuffer.cpp [hunk 8]
```diff
@@ -87,26 +107,48 @@ void GlStageBuffer::flushToGPU()
...
+}
+
+GLuint GlStageBuffer::getBufferId()
+{
+    return mGpuBuffer->getBufferId();
+}
+
+void GlStageBuffer::alignOffset()
+{
+
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlGpuBuffer.cpp:141]: -- empty space

File: src/renderer/gl_engine/tvgGlGpuBuffer.cpp [hunk 9]
```diff
@@ -87,26 +107,48 @@ void GlStageBuffer::flushToGPU()
...
+    uint32_t alignment = _getGpuBufferAlign();
+
+    if (mStageBuffer.count % alignment == 0) return;
+
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlGpuBuffer.cpp:145]: --empty space

File: src/renderer/gl_engine/tvgGlGpuBuffer.h [hunk 10]
```diff
@@ -34,31 +34,37 @@ class GlGpuBuffer
...
     void flushToGPU();
 
     void bind();
 
     void unbind();
 
+    GLuint getBufferId();
+private:
+    void alignOffset();
 private:
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlGpuBuffer.h:68]: --duplicated  private:

File: src/renderer/gl_engine/tvgGlRenderTask.cpp [hunk 11]
```diff
@@ -21,268 +21,62 @@
...
-            case VertexProperty::DataType::MATRIX: {
-                mProgram->setUniform4x4Value(property.second.propertyId, propertyVal.getCount(), propertyVal.getData());
-                break;
-            }
-        }
+    // setup attribute layout
+    for (uint32_t i = 0; i < mVertexLayout.count; i++) {
+        const auto &layout = mVertexLayout[i];
+        GL_CHECK(glEnableVertexAttribArray(layout.index));
+        GL_CHECK(glVertexAttribPointer(layout.index, layout.size, GL_FLOAT,
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlRenderTask.cpp:38]: one line please

File: src/renderer/gl_engine/tvgGlRenderTask.cpp [hunk 12]
```diff
@@ -21,268 +21,62 @@
...
-}
-
-
-void GlLinearGradientRenderTask::setStartPosition(float posX, float posY)
-{
-    if (mLocStartPos != -1)
-    {
-        PropertyInterface::setProperty(this, mLocStartPos, posX, posY);
+            GL_CHECK(glUniformBlockBinding(mProgram->getProgramId(), binding.location, binding.bindPoint));
+            GL_CHECK(glBindBufferRange(GL_UNIFORM_BUFFER, binding.bindPoint, binding.gBufferId,
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlRenderTask.cpp:54]: one line please

File: src/renderer/gl_engine/tvgGlRenderTask.h [hunk 13]
```diff
@@ -23,102 +23,77 @@
...
-    std::shared_ptr<GlProgram> mProgram;
-    std::map<int32_t, VertexProperty> mAttributePropertyBuffer;
-    std::map<int32_t, VertexProperty> mUniformPropertyBuffer;
-
-    int32_t mLocVertexAttribute = -1;
-    int32_t mLocTransform = -1;
+    uint32_t index;
+    uint32_t size;
+    uint32_t stride;
+    size_t   offset;
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlRenderTask.h:35]: size_t offset;

File: src/renderer/gl_engine/tvgGlRenderTask.h [hunk 14]
```diff
@@ -23,102 +23,77 @@
...
-    GlLinearGradientRenderTask();
-    void setStartPosition(float posX, float posY);
-    void setEndPosition(float posX, float posY);
+    GlBindingType type;
+    /**
+     * Binding point index.
+     *  Can be a uniform location for a texture
+     *  Can be a uniform buffer binding index for a uniform block
+     */
+    uint32_t        bindPoint = {};
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlRenderTask.h:53]: coding convention:
uint32_t bindPoint ...

File: src/renderer/gl_engine/tvgGlRenderTask.h [hunk 15]
```diff
@@ -23,102 +23,77 @@
...
 {
-public:
-    static std::shared_ptr<GlLinearGradientRenderTask> gen();
-    GlLinearGradientRenderTask();
-    void setStartPosition(float posX, float posY);
-    void setEndPosition(float posX, float posY);
+    GlBindingType type;
+    /**
+     * Binding point index.
+     *  Can be a uniform location for a texture
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlRenderTask.h:50]: align the space?
*__Can -> *_Can

File: src/renderer/gl_engine/tvgGlRenderTask.h [hunk 16]
```diff
@@ -23,102 +23,77 @@
...
+    uint32_t        bindPoint = {};
+    uint32_t        location = {};
+    GLuint          gBufferId = {};
+    uint32_t        bufferOffset = {};
+    uint32_t        bufferRange = {};
+
+    GlBindingResource() = default;
+
+    GlBindingResource(uint32_t index, uint32_t location, uint32_t bufferId, uint32_t offset, uint32_t range)
+        : type(GlBindingType::kUniformBuffer),
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlRenderTask.h:62]: oneline ->  : type(GlBindingType::kUniformBuffer), bindPoint(index), location(location) ........

File: src/renderer/gl_engine/tvgGlRenderer.cpp [hunk 17]
```diff
@@ -295,6 +300,10 @@ GlRenderer* GlRenderer::gen()
     return new GlRenderer();
 }
 
+GlRenderer::GlRenderer()
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlRenderer.cpp:303]: GlRenderer::GlRenderer() : mGpuBuffer(new GlStageBuffer), mPrograms(), mRenderTasks()

File: src/renderer/gl_engine/tvgGlRenderer.cpp [hunk 18]
```diff
@@ -309,30 +318,56 @@ GlRenderer::~GlRenderer()
...
+            0,
+            loc,
+            mGpuBuffer->getBufferId(),
+            mGpuBuffer->push(matrix, 16 * sizeof(float), true),
+            16 * sizeof(float),
+        });
+    }
+    // color 
+    {
+        float color[4] = {
```
Comment (hermet) [inline src/renderer/gl_engine/tvgGlRenderer.cpp:352]: less LOC vs readability...

**Review by hermet — APPROVED** [review]
Great. There are no significant issues in the code, although I'm leaving the major logic changes to you. Just a few minor convention nitpicks. I will push the patch in 3 days, and you can feel free to update it before that. Thanks.

File: src/renderer/gl_engine/tvgGlGpuBuffer.cpp (hunk 5 above)
Comment (RuiwenTang) [reply src/renderer/gl_engine/tvgGlGpuBuffer.cpp]: I changed these property initialization to single line. 
> BTW, I think thorvg really needs to define its own coding style, and the clang format configuration would be very helpful

Comment (hermet) [reply src/renderer/gl_engine/tvgGlGpuBuffer.cpp]: ok, thanks. let's see how we can enhance the policy around 1.0 release

**Pull request merged**